src/libraries/System.Net.Mail/src/System/Net/Mime/QEncodedStream.cs

```diff
@@ -221,38 +217,149 @@ public int EncodeBytes(byte[] buffer, int offset, int count)
+        }
+
+        private static int GetCodepointSize(string value, int i)
+        {
+            // specific encoding for CRLF
+            if (value[i] == '\r' && i + 1 < value.Length && value[i + 1] == '\n')
+            {
+                return 2;
+            }
+
+            if (char.IsSurrogate(value[i]) && i + 1 < value.Length && char.IsSurrogatePair(value[i], value[i + 1]))
+            {
+                return 2;
+            }
+
+            return 1;
+        }
+
+        private int AppendEncodedCodepoint(byte[] bytes, int count)
+        {
+            if (LineBreakNeeded(bytes, count))
+            {
+                WriteState.AppendCRLF(true);
+            }
+
+            // specific encoding for CRLF
+            if (IsCRLF(bytes, count))
+            {
+                AppendEncodedCRLF();
+                return 2;
+            }
+
+            for (int i = 0; i < count; ++i)
+            {
+                ApppendEncodedByte(bytes[i]);
+            }
+            return count;
+        }
+
+        private int ApppendEncodedByte(byte b)
+        {
+            if (b == ' ')
+            {
+                //spaces should be escaped as either '_' or '=20' and
+                //we have chosen '_' for parity with other email client
+                //behavior
+                WriteState.Append((byte)'_');
+            }
+            // RFC 2047 Section 5 part 3 also allows for !*+-/ but these arn't required in headers.
+            // Conservatively encode anything but letters or digits.
+            else if (IsAsciiLetterOrDigit((char)b))
+            {
+                // Just a regular printable ascii char.
+                WriteState.Append(b);
+            }
+            else
+            {
+                //append an = to indicate an encoded character
+                WriteState.Append((byte)'=');
+                //shift 4 to get the first four bytes only and look up the hex digit
+                WriteState.Append((byte)HexConverter.ToCharUpper(b >> 4));
+                //clear the first four bytes to get the last four and look up the hex digit
+                WriteState.Append((byte)HexConverter.ToCharUpper(b));
+            }
+            return 1;
+        }
+
+        private int AppendEncodedCRLF()
+        {
+            //the encoding for CRLF is =0D=0A
+            WriteState.Append((byte)'=', (byte)'0', (byte)'D', (byte)'=', (byte)'0', (byte)'A');
+            return 2;
+        }
+
+        private bool LineBreakNeeded(byte b)
+        {
+            // Fold if we're before a whitespace and encoding another character would be too long
+            return ((WriteState.CurrentLineLength + SizeOfFoldingCRLF + WriteState.FooterLength >= WriteState.MaxLineLength)
+                && (b == ' ' || b == '\t' || b == '\r' || b == '\n'))
+                // Or just adding the footer would be too long.
+                || (WriteState.CurrentLineLength + _writeState.FooterLength >= WriteState.MaxLineLength);
```

Comment:
Please, consider extracting operands into local variables with meaningful names to improve readability.

Reply:
Fixed